components/job-orchestration/job_orchestration/scheduler/search/search_scheduler.py

```diff
@@ -291,28 +312,29 @@ def handle_pending_search_jobs(
             else:
                 pending_jobs.append(new_search_job)
             active_jobs[job_id] = new_search_job
-        db_conn.commit()
 
-    for job in pending_jobs:
-        job_id = job.id
-
-        if (
-            job.search_config.network_address is None
-            and len(job.remaining_archives_for_search) > num_archives_to_search_per_sub_job
-        ):
-            archives_for_search = job.remaining_archives_for_search[
-                :num_archives_to_search_per_sub_job
-            ]
-            job.remaining_archives_for_search = job.remaining_archives_for_search[
-                num_archives_to_search_per_sub_job:
-            ]
-        else:
-            archives_for_search = job.remaining_archives_for_search
-            job.remaining_archives_for_search = []
+        for job in pending_jobs:
+            job_id = job.id
+
+            if (
+                job.search_config.network_address is None
+                and len(job.remaining_archives_for_search) > num_archives_to_search_per_sub_job
+            ):
+                archives_for_search = job.remaining_archives_for_search[
+                    :num_archives_to_search_per_sub_job
+                ]
+                job.remaining_archives_for_search = job.remaining_archives_for_search[
+                    num_archives_to_search_per_sub_job:
+                ]
+            else:
+                archives_for_search = job.remaining_archives_for_search
+                job.remaining_archives_for_search = []
 
-        dispatch_search_job(job, archives_for_search, results_cache_uri)
-        logger.info(f"Dispatched job {job_id} with {len(archives_for_search)} archives to search.")
-        set_job_status(db_conn, job_id, SearchJobStatus.RUNNING, SearchJobStatus.PENDING)
+            dispatch_search_job(job, archives_for_search, results_cache_uri)
+            logger.info(
+                f"Dispatched job {job_id} with {len(archives_for_search)} archives to search."
+            )
+            set_job_status(db_conn, job_id, SearchJobStatus.RUNNING, SearchJobStatus.PENDING)
```

Comment:
Currently, it seems like database updates can fail meaning the scheduler's in-memory state will not match the state of the database (this was true previously as well, but we wouldn't continue running if the database update failed). Have we reasoned about all possible scenarios? If so, the scenarios would be good to call out in the PR description.

Reply:
I think the short answer is that there are some scenarios that don't work in this implementation that didn't work in the previous implementation, but sometimes in a slightly different way.

I'd say we maybe add some comments or something that calls out that failure semantics surrounding set_job_status are not well defined under the conditions of the database being unreachable, and we handle those types of failures in a separate PR.

Reply:
Can you enumerate them quickly? Since we're no longer fail-stop, it will be harder to detect these failures, so I think it's important to know what they are.

Reply:
Cases I can think of are
1) everything works until the final update to mark the job completed at which point the database becomes unreachable
This is identical to before in that either the job has succeeded or failed + all results are in the cache, but the job gets stuck in RUNNING status in the db forever
2) the initial update to mark the job as running fails because the database is unreachable
Results will be uploaded to the results cache, but the job will be stuck in PENDING status in the database unless it is ever cancelled or the database comes back up and we pick up the job a second time
3) the job is marked cancelling and the final update to mark the job cancelled fails because the database is unreachable 
This actually works correctly, because whenever the database comes back up we will pick up jobs in CANCELLING state and mark them CANCELLED
4) we pick up the job in the database but don't retrieve any matching archives because the database is unreachable 
In this case we will terminate the query with "no matching archives", but the job will be stuck in PENDING status in the database unless it is ever cancelled or the database comes back up and we pick up the job a second time

Reply:
Thanks.

Can you do two things:

1. Add these failure scenarios to the PR description.
2. Add this docstring at the top of the `search_scheduler` module (you can replace the shebang):

```
"""
A scheduler for scheduling search jobs in the CLP package.

NOTE: This scheduler currently only has partial handling for failures of the database. Specifically,
in the event that the database is unreachable, the scheduler will continue running as if reads from
the database return no records and writes to the database always fail. Failed writes are currently
silently ignored, in which case the state of the database won't match the scheduler's internal
state. If the database comes back up, the scheduler will eventually reconnect to it and reads/writes
will function as normal again. However, the mismatched state may lead to unexpected behaviour like
jobs seemingly being stuck in the "RUNNING" state or jobs being repeated, which in turn will create
duplicated search results in the results cache, possibly long after the results had already been
cleared from the cache. Unfortunately, these effects will require manual intervention to clean-up.
TODO Address this limitation.
"""
```

Reply:
@gibber9809